clang/lib/DPCT/ASTTraversal.cpp

```diff
@@ -11894,11 +11897,62 @@ void CooperativeGroupsFunctionRule::registerMatcher(MatchFinder &MF) {
                                                hasAttr(attr::CUDAGlobal))))))
           .bind("FuncCall"),
       this);
+  MF.addMatcher(
+      declRefExpr(
+          hasAncestor(
+              implicitCastExpr(
+                  hasImplicitDestinationType(qualType(hasCanonicalType(
+                      recordType(hasDeclaration(cxxRecordDecl(hasName(
+                          "cooperative_groups::__v1::thread_group"))))))))))
+          .bind("declRef"),
+      this);
 }
 
 void CooperativeGroupsFunctionRule::runRule(
     const MatchFinder::MatchResult &Result) {
   const CallExpr *CE = getNodeAsType<CallExpr>(Result, "FuncCall");
+  const DeclRefExpr *DR = getNodeAsType<DeclRefExpr>(Result, "declRef");
+  const SourceManager &SM = DpctGlobalInfo::getSourceManager();
+  if (DR) {
+    std::string DREName = DR->getType().getCanonicalType().getAsString();
+    std::string ReplacedStr = "";
+    if (DpctGlobalInfo::getAssumedNDRangeDim() == 1) {
+      auto FD = DpctGlobalInfo::getParentFunction(DR);
+      if (!FD)
+        return;
+      auto DFI = DeviceFunctionDecl::LinkRedecls(FD);
+      auto Index = DpctGlobalInfo::getCudaKernelDimDFIIndexThenInc();
+      ReplacedStr = "<{{NEEDREPLACEG" + std::to_string(Index) + "}}>";
+      DpctGlobalInfo::insertCudaKernelDimDFIMap(Index, DFI);
+    } else {
+      ReplacedStr = "<3>";
+    }
+
+    if (DREName.find("cooperative_groups::__v1::thread_block_tile<32>") !=
+        std::string::npos) {
+      ReplacedStr = "sycl::sub_group";
+    } else if (DREName == "class cooperative_groups::__v1::thread_block") {
+      ReplacedStr = "sycl::group" + ReplacedStr;
+    } else {
+      ReplacedStr = "dpct::experimental::logical_group" + ReplacedStr;
+    }
+
+    ReplacedStr = MapNames::getDpctNamespace() + "item_group<" + ReplacedStr +
+                  ">(" + DR->getNameInfo().getAsString() + ", " +
+                  DpctGlobalInfo::getItem(DR) + ")";
+    SourceLocation Begin = DR->getBeginLoc();
+    SourceLocation End = DR->getEndLoc();
+    End = End.getLocWithOffset(Lexer::MeasureTokenLength(
+        End, SM, DpctGlobalInfo::getContext().getLangOpts()));
+    if (End.isMacroID())
+      return;
+    if (Begin.isMacroID() || End.isMacroID())
+      return;
+    emplaceTransformation(
+        new ReplaceText(Begin, End.getRawEncoding() - Begin.getRawEncoding(),
```

Comment:
Advice to add an assert() to detect overflow of "End.getRawEncoding() - Begin.getRawEncoding()".

Reply:
Updated